### ISSUE TITLE
Inference speed does not improve after using flash attention sdpa operator

### ISSUE THREAD

**Issue by bohyunshin** [huggingface/optimum#1341]
### System Info

```shell
## version info
* optimum==1.12.0
* transformers==4.32.1

$ cat requirements.txt
aiohttp==3.8.5
aiosignal==1.3.1
async-timeout==4.0.3
attrs==23.1.0
certifi==2023.7.22
charset-normalizer==3.2.0
cmake==3.27.2
coloredlogs==15.0.1
datasets==2.14.4
dill==0.3.7
filelock==3.12.3
frozenlist==1.4.0
fsspec==2023.9.0
huggingface-hub==0.16.4
humanfriendly==10.0
idna==3.4
Jinja2==3.1.2
lit==16.0.6
MarkupSafe==2.1.3
mpmath==1.3.0
multidict==6.0.4
multiprocess==0.70.15
networkx==3.1
numpy==1.25.2
nvidia-cublas-cu11==11.10.3.66
nvidia-cuda-cupti-cu11==11.7.101
nvidia-cuda-nvrtc-cu11==11.7.99
nvidia-cuda-runtime-cu11==11.7.99
nvidia-cudnn-cu11==8.5.0.96
nvidia-cufft-cu11==10.9.0.58
nvidia-curand-cu11==10.2.10.91
nvidia-cusolver-cu11==11.4.0.1
nvidia-cusparse-cu11==11.7.4.91
nvidia-nccl-cu11==2.14.3
nvidia-nvtx-cu11==11.7.91
optimum==1.12.0
packaging==23.1
pandas==2.1.0
protobuf==4.24.2
pyarrow==13.0.0
pyspark==3.2.1
python-dateutil==2.8.2
pytz==2023.3
PyYAML==6.0.1
regex==2023.8.8
requests==2.31.0
safetensors==0.3.3
sentencepiece==0.1.99
six==1.16.0
sympy==1.12
tokenizers==0.13.3
torch==2.0.1
tqdm==4.66.1
transformers==4.32.1
triton==2.0.0
typing_extensions==4.7.1
tzdata==2023.3
urllib3==2.0.4
xxhash==3.3.0
yarl==1.9.2


## system info
* centos linux ver. 7
```bash
$ cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

## gpu info
* A100 40GB MEMORY 2EA

## python version
* python==3.9.16
```bash
$ python -m venv .venv python=3.9
$ source .venv/bin/activate
```
```


### Who can help?

_No response_

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [X] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction (minimal, reproducible, runnable)

## set up virtual env
```bash
$ python -m venv .venv python=3.9
$ source .venv/bin/activate
$ pip install -r requirements.txt # see the above requirements.txt
```
## run the following python script
* Referenced from [official docs](https://github.com/huggingface/transformers/blob/main/docs/source/en/perf_infer_gpu_one.md)
```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer
import time

tokenizer = AutoTokenizer.from_pretrained("facebook/opt-350m")
model = AutoModelForCausalLM.from_pretrained("facebook/opt-350m", torch_dtype=torch.float16).to("cuda")
# convert the model to BetterTransformer
model.to_bettertransformer()

input_text = "Hello my dog is cute and"
inputs = tokenizer(input_text, return_tensors="pt").to("cuda")

s = time.time()
with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    outputs = model.generate(**inputs, max_length=2000)
print(f"Inference time for flash attention kernel: {time.time()-s}")


model.reverse_bettertransformer()
s = time.time()
with torch.backends.cuda.sdp_kernel(enable_flash=False, enable_math=True, enable_mem_efficient=False):
    outputs = model.generate(**inputs, max_length=2000)
print(f"Inference time for ordinary sdpa attention kernel: {time.time()-s}")
```

## results
```
The BetterTransformer implementation does not support padding during training, as the fused kernels do not support attention masks. Beware that passing padded batched data during training may result in unexpected outputs. Please refer to https://huggingface.co/docs/optimum/bettertransformer/overview for more details.
Inference time for flash attention kernel: 0.74764084815979
Inference time for ordinary sdpa attention kernel: 0.16220331192016602
```

### Expected behavior

Expected shorter inference time when using flash attention than using ordinary sdpa from pytorch 2.0, but got longer inference time with flash attention backend.
I tried with another prompt which results in longer generated text, but result is almost identical, i.e., almost similar inference time between flash attention and sdpa from pytorch.

Anything wrong from the above code? Thanks!

## Reference
* https://github.com/huggingface/transformers/blob/main/docs/source/en/perf_infer_gpu_one.md
* https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention

**Comment by fxmarty** [comment]
@bohyunshin Thank you for the report. I can not reproduce the issue.

Could you try to run `CUDA_VISIBLE_DEVICES=0 python benchmark_bettertransformer.py --model-name facebook/opt-350m --is_decoder --use-half --use-cuda --batch-size 1 --num-batches 10` with [this script](https://github.com/huggingface/optimum/blob/main/tests/benchmark/benchmark_bettertransformer.py) and report the result?

On an A100 I am getting the following result:
|   num_batches |    batch_size |    seq_len |  is cuda   |  is half   |  use mask   |    pad percentage |    Latency eager (ms) |    Latency BT (ms) |    Speedup (%) |    Mem eager (MB) |    Mem BT (MB) |    Mem saved (%) |
|--------------:|--------------:|-----------:|:-----------|:-----------|:------------|------------------:|----------------------:|-------------------:|---------------:|------------------:|---------------:|-----------------:|
|            10 |             1 |        256 | True       | True       | False       |                 0 |               1266.57 |             1072.6 |         18.084 |           749.116 |        749.113 |                0 |

Several issues there could be in your script:
* there is no warmup
* you measure time with `time.time()` which is unreliable, especially given that PyTorch cuda kernel invocation is asynchronous.
* you do not clear cuda cache between the two benchmarks.
* you measure only a single generate call.
* You do not set `eos_token_id` to None, and do not set a `min_new_tokens`, so in the event (very highly unlikely given that it is tested) that the two implementations do not match, we are actually not sure 2000 tokens are generated.

It could also simply be due to the GPU you are using, for which flash may not be performant.

**Comment by bohyunshin** [comment]
@fxmarty  Thank you for pointing out potential issues in my script. There are lots of things to consider!

After running the shared script, I got following results.

|   num_batches |    batch_size |    seq_len |  is cuda   |  is half   |  use mask   |    pad percentage |    Latency eager (ms) |    Latency BT (ms) |    Speedup (%) |    Mem eager (MB) |    Mem BT (MB) |    Mem saved (%) |
|--------------:|--------------:|-----------:|:-----------|:-----------|:------------|------------------:|----------------------:|-------------------:|---------------:|------------------:|---------------:|-----------------:|
|            10 |             1 |        256 | True       | True       | False       |                 0 |               1400.67 |            1150.03 |         21.795 |           749.116 |        749.113 |                0 |

So, similar to your results, I got 21% latency improvement and no improvement on `bettertransformer`.
But still curious why there are not any improvements on memory performance, because flash attention is known for their reduction on memory complexity. ($O(n^2) \rightarrow O(n)$)
Probably, what you pointed out could be the reasons why I did not get improvement on the results.

Thank you for your help!

**Comment by bohyunshin** [comment]
@fxmarty 
By the way, may I ask you some more questions?

Could you explain some more detail about what you said?
> It could also simply be due to the GPU you are using, for which flash may not be performant.

Also, the script which you shared do not apply flash attention only because several other options are enabled in [this code](https://github.com/huggingface/optimum/blob/main/tests/benchmark/benchmark_bettertransformer.py#L269-L270). So, if I want to purely apply flash attention, I should modify code as below right?

```python
with torch.backends.cuda.sdp_kernel(
        enable_flash=True, enable_math=False, enable_mem_efficient=False
    ):
```

Finally, this is another question which is not related to this issue.
In [this official optimum docs](https://huggingface.co/docs/optimum/bettertransformer/overview), the supported models for `bettertransformer` do not include llama. However, in [this official transformer docs](https://github.com/huggingface/transformers/blob/main/docs/source/en/perf_infer_gpu_one.md#decoder-models), llama is included as supported decoder models. I'm somewhat confused which is correct. Could you tell me which docs is correct, or I'm looking at wrong docs?

Thank you again for your kind help!

**Comment by fxmarty** [comment]
Hi @bohyunshin thank you for the interesting questions!

>  But still curious 🧐 why there are not any improvements on memory performance, because flash attention is known for their reduction on memory complexity. ($O(n^2) \rightarrow O(n)$)

Good question, I am not 100% sure why. I had the same result when benchmarking more thoroughly for the blog post https://pytorch.org/blog/out-of-the-box-acceleration/. Peak memory usage is measured through `torch.cuda.max_memory_allocated` which could be not 100% reliable. It could be that the memory bottleneck during inference is not in the attention but somewhere else in the model (why would it though? maybe during the prefill?).

A answer that seem more plausible to me is: Flash Attention allows to avoid GPU global memory <-> local/cache/SM buffers communications. Thus it does not really allow to lower the peak global memory usage during the forward, but it helps during the backward as some intermediate activations then do not need to be saved (as FA implements its own backward of the attention).

> It could also simply be due to the GPU you are using, for which flash may not be performant.

Flash Attention v1 supports Turing, Ampere, Ada, or Hopper GPUs: https://github.com/Dao-AILab/flash-attention/tree/v1.0.9 It may be though that on different GPUs, Flash brings more or less speedup.

> Third, the benchmark measures performance of the method based on latency and memory metric, but I think tokens per second could be more accurate for objective metric.

Yes it would be a relevant metric as well. It is basically latency vs throughput. Both are relevant metrics, depending on your use case.

> Because get_batch function in [this code](https://github.com/huggingface/optimum/blob/main/tests/benchmark/benchmark_bettertransformer.py#L77) generate random tensors without specifiying any seed numbers, the generated batch for bt model and not bt model will be different and as a result, the generated text from model will be different. In this situation, both models could generate different length of text

Indeed this should be fixed, good catch! Although here it does not really matter, the only difference will be the amount of padding, which neither of transformers vanilla nor SDPA exploit (for decoder models), since we use `min_new_tokens = max_new_tokens` & `eos_token_id=None` so that exactly the number of `--max_token` are generated.

> In [this official optimum docs](https://huggingface.co/docs/optimum/bettertransformer/overview), the supported models for bettertransformer do not include llama. However, in [this official transformer docs](https://github.com/huggingface/transformers/blob/main/docs/source/en/perf_infer_gpu_one.md#decoder-models), llama is included as supported decoder models. I'm somewhat confused which is correct. Could you tell me which docs is correct, or I'm looking at wrong docs?

It should be there. I will fix, thank you!

**Comment by bohyunshin** [comment]
Hi @fxmarty , thank you for sharing interesting answers!

> Good question, I am not 100% sure why. I had the same result when benchmarking more thoroughly for the blog post https://pytorch.org/blog/out-of-the-box-acceleration/. Peak memory usage is measured through torch.cuda.max_memory_allocated which could be not 100% reliable. It could be that the memory bottleneck during inference is not in the attention but somewhere else in the model (why would it though? maybe during the prefill?).
An answer that seem more plausible to me is: Flash Attention allows to avoid GPU global memory <-> local/cache/SM buffers communications. Thus it does not really allow to lower the peak global memory usage during the forward, but it helps during the backward as some intermediate activations then do not need to be saved (as FA implements its own backward of the attention).

I do not really have a deep understanding of FA 😅, but seeing the benchmark memory code in [official FA repo](https://github.com/Dao-AILab/flash-attention/blob/v1.0.9/flash_attn/utils/benchmark.py#L136), what you wrote for [benchmark code](https://github.com/huggingface/optimum/blob/main/tests/benchmark/benchmark_bettertransformer.py#L108) is almost identical to official FA benchmark code without `torch.cud.empty_cache()` [here](https://github.com/Dao-AILab/flash-attention/blob/v1.0.9/flash_attn/utils/benchmark.py#L145). May be you excluded it for some reasons right? (noted in [here](https://github.com/huggingface/optimum/blob/main/tests/benchmark/benchmark_bettertransformer.py#L115))

> Flash Attention v1 supports Turing, Ampere, Ada, or Hopper GPUs: https://github.com/Dao-AILab/flash-attention/tree/v1.0.9 It may be though that on different GPUs, Flash brings more or less speedup.

I did not recognize it, thank you for pointing out!

> Indeed this should be fixed, good catch! Although here it does not really matter, the only difference will be the amount of padding, which neither of transformers vanilla nor SDPA exploit (for decoder models), since we use min_new_tokens = max_new_tokens & eos_token_id=None so that exactly the number of --max_token are generated.

I did not recognize this neither. This looks like a good trick for llm forcely generate same amount of tokens !

> It should be there. I will fix, thank you!

May be I could open a simple, minimum PR to include llama for supported models in official docs !

**Comment by fxmarty** [comment]
@bohyunshin Yes sure feel free to open a PR!

Okay so all my hypothesis for the no-memory-reduction are probably wrong, maybe it is simply due to too short sequence length: https://github.com/Dao-AILab/flash-attention/issues/403#issuecomment-1658844143

**Comment by bohyunshin** [comment]
Hi @fxmarty , thank you for sharing related github issue!
Seems that flash attention optimizes attention related operations not other operations like MLP, so when the generated sequence for benchmark is too short, it is probable that memory improvement could not happen. That's interesting to know.
It will be interesting to see the changes of memory savings with extending the length of sequences :)
